File: algorithmic-toolbox/week2/solutions/5_fibonacci_number_again/python/fibonacci_huge.py
# 1) Find Pisano Period Length of modulo m; pisano_length
# 2) r = n % pisano_length
# 3) F(r) % m


def get_fibonacci_period_length(m):
    """
    Given modulo `m`, return the Pisano Period length.

    The Pisano period is the sequence of remainers r returned from taking
    modulo m of Fibonacci(0)...Fibonacci(n).
    """
    assert m >= 2
    pisano_length = 0
    prev, current = 0, 1
    while True:
        current, prev = (current + prev) % m, current
        pisano_length += 1
        if current == 1 and prev == 0:
            return pisano_length


def fibonacci(n: int):
    if n <= 1:
        return n
    prev, current = 0, 1
    for i in range(1, n):
        current, prev = current + prev, current

    return current


def fibonacci_mod(n: int, m: int):
    if n <= 1:
        return n
    prev, current = 0, 1
    for i in range(1, n):
        current, prev = (current + prev) % m, current

    return current


def fibonacci_mod_m(n, m):
    pisano_length = get_fibonacci_period_length(m)
    remainder = n % pisano_length
    # return fibonacci(remainder)
    return fibonacci_mod(remainder, m)


if __name__ == "__main__":
    n, m = map(int, input().split())
    print(fibonacci_mod_m(n, m))

File: algorithmic-toolbox/week2/solutions/5_fibonacci_number_again/cpp/fibonacci_huge.cpp
#include <iostream>
#include <cassert>

long long fibonacci_pisano_period_length(long long m)
{
    long long pisano_length = 0;
    long long prev = 0;
    long long current = 1;
    long long tmp_swap = 0;
    while (true)
    {
        tmp_swap = current;
        current = (current + prev) % m;
        prev = tmp_swap;
        ++pisano_length;
        if (current == 1 && prev == 0)
        {
            return pisano_length;
        }
    }
}

long long fibonacci(long long n)
{
    if (n <= 1)
    {
        return n;
    }
    long long prev = 0;
    long long current = 1;
    long long swap = 0;
    for (long long i = 1; i < n; ++i)
    {
        swap = current;
        current = current + prev;
        prev = swap;
    }
    return current;
}

long long fibonacci_mod(long long n, long long m)
{
    if (n <= 1)
    {
        return n;
    }
    long long prev = 0;
    long long current = 1;
    long long swap = 1;
    for (long long i = 1; i < n; ++i)
    {
        swap = current;
        current = (current + prev) % m;
        prev = swap;
    }
    return current;
}

long long fibonacci_mod_m(long long n, long long m)
{
    long long pisano_length = fibonacci_pisano_period_length(m);
    long long remainder = n % pisano_length;
    return fibonacci_mod(remainder, m);
}

void test_solution()
{
    assert(fibonacci(10) == 55);
    assert(fibonacci_pisano_period_length(2) == 3);
    assert(fibonacci_pisano_period_length(3) == 8);
    assert(fibonacci_mod_m(239, 1000) == 161);
    assert(fibonacci_mod_m(2816213588, 239) == 151);
}

int main()
{
    // test_solution();

    long long n, m;
    std::cin >> n >> m;
    std::cout << fibonacci_mod_m(n, m) << '\n';
}
